include/onnxruntime/core/framework/data_types_internal.h

```diff
@@ -286,118 +280,139 @@ class CallableDispatchableRetHelper {
-// Return type of Fn must return type convertible to Ret
-// The value of the return type will be the return value
-// of the function for type T which was specified for execution.
-template <class Ret, template <typename> class Fn, typename... Types>
-class MLTypeCallDispatcherRet {
-  int32_t dt_type_;
-
- public:
-  explicit MLTypeCallDispatcherRet(int32_t dt_type) noexcept : dt_type_(dt_type) {}
-
-  template <typename... Args>
-  Ret Invoke(Args&&... args) const {
-    mltype_dispatcher_internal::CallableDispatchableRetHelper<Ret> helper(dt_type_);
-    int results[] = {0, helper.template Invoke<Types>(Fn<Types>(), std::forward<Args>(args)...)...};
-    ORT_UNUSED_PARAMETER(results);
-    return helper.Get();
-  }
-
-  template <class UnsupportedPolicy, typename... Args>
-  Ret InvokeWithUnsupportedPolicy(Args&&... args) const {
-    mltype_dispatcher_internal::CallableDispatchableRetHelper<Ret, UnsupportedPolicy> helper(dt_type_);
-    int results[] = {0, helper.template Invoke<Types>(Fn<Types>(), std::forward<Args>(args)...)...};
-    ORT_UNUSED_PARAMETER(results);
-    return helper.Get();
-  }
-};
 
-// Version of MLTypeCallDispatcher that takes supported types as class-level template parameters.
-// This enables easier use with type list representations of the supported types.
-// The invocation-related template parameters like Fn move to the individual Invoke() methods.
-// TODO consolidate this with the other MLTypeCallDispatcher classes
-// can add additional methods to cover their usages, but need to update call sites
-template <typename... Types>
-class MLTypeCallDispatcher2 {
-  static_assert(boost::mp11::mp_is_set<TypeList<Types...>>::value,
-                "MLTypeCallDispatcher requires a set of unique types.");
-
-  int32_t dt_type_;
-
- public:
-  explicit MLTypeCallDispatcher2(int32_t dt_type) noexcept : dt_type_(dt_type) {}
+  /**
+   * Invokes Fn<..., T> with leading template arguments and the specified arguments.
+   *
+   * @tparam Fn The function object template.
+   * @tparam LeadingTemplateArgTypeList A type list of the leading template arguments.
+   * @tparam Args The argument types.
+   */
+  template <template <typename...> class Fn, typename LeadingTemplateArgTypeList, typename... Args>
+  void InvokeWithLeadingTemplateArgs(Args&&... args) const {
+    static_assert(
+        boost::mp11::mp_is_list<LeadingTemplateArgTypeList>::value,
+        "LeadingTemplateArgTypeList must be a type list (e.g., onnxruntime::TypeList<...>).");
 
-  template <template <typename> class Fn, typename... Args>
-  void Invoke(Args&&... args) const {
     mltype_dispatcher_internal::CallableDispatchableHelper helper(dt_type_);
 
     static_cast<void>(std::array<int, sizeof...(Types)>{
-        helper.template Invoke<Types>(Fn<Types>(), std::forward<Args>(args)...)...});
+        helper.template Invoke<Types>(
+            boost::mp11::mp_apply<Fn, boost::mp11::mp_push_back<LeadingTemplateArgTypeList, Types>>(),
```

Comment:
> boost::mp11::mp_push_back<LeadingTemplateArgTypeList, Types> [](start = 37, length = 61)

Why not simply SupportedTypesList?

Reply:
Okay, let's add a comment here.

---
In reply to: [574931521](https://github.com/microsoft/onnxruntime/pull/6651#discussion_r574931521) [](ancestors = 574931521)